### PR TITLE
Make Scripting parser utility compile under XE2

### PR DESCRIPTION
**Pull request by reyandme**
implementation of #247

To compile project you also need KaM_Remake.inc, but I did not find any other way, then add it locally it to the project folder. I did not add it to this PR, because there should be a better way.

### CODE REVIEW

File: src/KM_Utils.pas [hunk 1]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+  Result := -1;
+  for I := 1 to Length(aStr) do
+    if StartsStr(aSubStr, StrSubstring(aStr, I-1)) then
+      Result := I - 1;
+end;
+
+
+function StrStartsWith(aStr, aSubStr: String): Boolean;
+begin
+  Result := StartsStr(aSubStr, aStr);
```
Comment (Kromster80) [inline src/KM_Utils.pas]: Why not just replace StrStartsWith with StartsStr everywhere in code?

File: src/KM_Utils.pas [hunk 2]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+
+function StrSubstring(aStr: String; aFrom: Integer): String;
+begin
+  Result := Copy(aStr, aFrom + 1, Length(aStr));
+end;
+
+
+function StrSubstring(aStr: String; aFrom, aLength: Integer): String;
+begin
+  Result := Copy(aStr, aFrom + 1, aLength);
```
Comment (Kromster80) [inline src/KM_Utils.pas]: Why not just replace StrSubstring with Copy everywhere in code?

File: src/KM_Utils.pas [hunk 3]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
 
 
+{
+String functions
+These function are replacements for String functions introduced after XE2 (XE5 probably)
+Names are the same as in new Delphi versions, but with 'Str' prefix
+}
+function StrIndexOf(aStr, aSubStr: String): Integer;
+begin
+  Result := AnsiPos(aSubStr, aStr) - 1;
```
Comment (Kromster80) [inline src/KM_Utils.pas]: Why not just replace StrIndexOf with Pos everywhere in code?

File: src/KM_Utils.pas [hunk 4]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+
+function StrStartsWith(aStr, aSubStr: String): Boolean;
+begin
+  Result := StartsStr(aSubStr, aStr);
+end;
+
+
+function StrSubstring(aStr: String; aFrom: Integer): String;
+begin
+  Result := Copy(aStr, aFrom + 1, Length(aStr));
```
Comment (Kromster80) [inline src/KM_Utils.pas]: Why not just replace StrSubstring with RightStr everywhere in code?

File: src/KM_Utils.pas [hunk 5]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+
+function StrSubstring(aStr: String; aFrom, aLength: Integer): String;
+begin
+  Result := Copy(aStr, aFrom + 1, aLength);
+end;
+
+
+function StrContains(aStr, aSubStr: String): Boolean;
+begin
+  Result := StrIndexOf(aStr, aSubStr) <> -1;
```
Comment (Kromster80) [inline src/KM_Utils.pas]: Why not just replace StrContains with Pos() <> 0 everywhere in code?

File: src/KM_Utils.pas [hunk 6]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+begin
+  StrArray := SplitString(aStr, aDelimiters);
+  SetLength(Result, Length(StrArray));
+  for I := Low(StrArray) to High(StrArray) do
+    Result[I] := StrArray[I];
+end;
+
+
+function StrSplit(aStr, aDelimiters: String): TStrings;
+var StrArray: TStringDynArray;
```
Comment (Kromster80) [inline src/KM_Utils.pas:812]: It's bad practice to create object (TStringList) inside and return it as parent class (TStrings). Do we really need it this way? Better to pass TStringList from outside in a parameter.

File: src/KM_Utils.pas [hunk 7]
```diff
@@ -717,4 +732,93 @@ function GetMultiplicator(aShift: TShiftState): Word;
...
+      end;
+    end;
+    if not Found then
+      Break;
+  end;
+  Result := Copy(aStr, 1, I);
+end;
+
+
+function StrSplitToStrArray(aStr, aDelimiters: String): TStringArray;
```
Comment (Kromster80) [inline src/KM_Utils.pas]: This function seems to be unused.

**Review by Kromster80 — CHANGES_REQUESTED** [review]
Please review the comments

**Comment by reyandme** [comment]
I will add these @Kromster80 comments as Todo to code for future fixing then.

**Comment by Kromster80** [comment]
Merged with heavy heart. Hopefully "todos" will be resolved soon.

**Comment by lewinjh** [comment]
@reyandme I agree with @Kromster80, reinventing our own versions of standard functions is a very bad idea. The compiler implementations will be significantly more efficient (often hand optimised assembly), and it makes the code less portable. I'd appreciate it if you can fix these todos soon.

**Pull request merged**